File: Pastura/PasturaTests/Engine/ErrorReadabilityTests.swift
import Foundation
import Testing

@testable import Pastura

@Suite(.timeLimit(.minutes(1)))
struct ErrorReadabilityTests {
  // MARK: - Helper direct behavior

  @Test func prefersLocalizedErrorDescription() {
    struct LocalizedFake: LocalizedError {
      var errorDescription: String? { "human readable message" }
    }
    let result = readableDescription(LocalizedFake())
    #expect(result.contains("human readable"))
  }

  @Test func fallsBackWhenLocalizedErrorReturnsNil() {
    struct LocalizedNil: LocalizedError {
      var errorDescription: String? { nil }
    }
    // Falls through to String(describing:), which prints the type name for an
    // empty struct — we only assert the helper doesn't crash and returns
    // something non-empty.
    let result = readableDescription(LocalizedNil())
    #expect(!result.isEmpty)
  }

  @Test func fallsBackToStringDescribingForPlainError() {
    enum PlainError: Error { case boom }
    let result = readableDescription(PlainError.boom)
    #expect(result.contains("boom"))
  }

  // MARK: - Wrap-chain semantics at the Engine boundary

  // These tests assert that the helper preserves the inner error's meaningful
  // text when Engine wraps foreign errors into SimulationError. They inspect
  // the wrapped case's associated-value String directly rather than reading
  // `wrapped.localizedDescription` — that final surfacing step depends on
  // SimulationError adopting LocalizedError (landed in a subsequent commit)
  // and is exercised via the per-enum `errorDescription` tests. Keeping this
  // test focused on the helper's contribution makes it immune to the
  // conformance rollout order.

  @Test func wrapChainCarriesInnerLLMErrorText() {
    let inner = LLMError.generationFailed(description: "connection timeout")
    let wrapped = SimulationError.llmGenerationFailed(
      description: readableDescription(inner))
    guard case .llmGenerationFailed(let desc) = wrapped else {
      Issue.record("expected .llmGenerationFailed case")
      return
    }
    #expect(desc.contains("connection timeout"))
  }

  @Test func wrapChainCarriesLocalizedErrorText() {
    enum ValidatorThrown: LocalizedError {
      case missingSource(String)
      var errorDescription: String? {
        switch self {
        case .missingSource(let name): return "source '\(name)' not found"
        }
      }
    }
    let inner = ValidatorThrown.missingSource("topics")
    let wrapped = SimulationError.scenarioValidationFailed(
      readableDescription(inner))
    guard case .scenarioValidationFailed(let desc) = wrapped else {
      Issue.record("expected .scenarioValidationFailed case")
      return
    }
    #expect(desc.contains("source 'topics' not found"))
    // Prefers LocalizedError.errorDescription over the raw enum form —
    // guards against a regression to `"\(error)"` stringification.
    #expect(!desc.contains("missingSource("))
  }
}

File: shared/engine/src/commonTest/kotlin/com/pastura/engine/ErrorReadabilityTests.kt
package com.pastura.engine

import com.pastura.models.SimulationError
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * Parity spec for [readableDescription] — mirrors the intent of the Swift
 * `ErrorReadabilityTests` wrap-chain cases against the Kotlin
 * `message ?: toString()` mapping. The message-present cases go red if the
 * `?:` fallback collapses to a `toString()`-always impl (which would leak the
 * class name).
 */
class ErrorReadabilityTests {

    @Test
    fun prefersMessageWhenPresent() {
        val result = readableDescription(RuntimeException("boom"))
        assertEquals("boom", result)
        // Regression guard: a `toString()`-always impl would yield
        // "...RuntimeException: boom", leaking the class name.
        assertTrue(!result.contains("Exception"))
    }

    @Test
    fun fallsBackToToStringWhenMessageIsNull() {
        val e = RuntimeException()
        val result = readableDescription(e)
        // The exact class-name string is platform-specific; assert only that
        // the `?: toString()` branch produced non-empty text equal to toString().
        assertTrue(result.isNotEmpty())
        assertEquals(e.toString(), result)
    }

    @Test
    fun wrapChainCarriesInnerSimulationErrorText() {
        val wrapped = SimulationException(SimulationError.LlmGenerationFailed("connection timeout"))
        val result = readableDescription(wrapped)
        assertTrue(result.contains("connection timeout"))
        // Regression guard: preferring `.message` strips the sealed-subtype
        // wrapper; a raw `toString()` would leak it.
        assertTrue(!result.contains("LlmGenerationFailed("))
    }
}
